### ISSUE TITLE
Question: Correct way to use ActivatorUtilities.CreateInstance with DI resolved params and provided params.

### ISSUE THREAD

**Issue by mhsimkin** [aspnet/Extensions#949]
I have a question on ActivatorUtilities.CreateInstance.  I’m making this call:

`ActivatorUtilities.CreateInstance<RabbitMqConnection>(_serviceProvider, connectionConfiguration, cancellationToken, notifications);`

The constructor on RabbitMqConnecton is defined as: 

```
public RabbitMqConnection(System.IServiceProvider serviceProvider, ILogger<RabbitMqConnection> logger,
            ConnectionConfiguration connectionConfiguration, CancellationToken cancellationToken,
            IMessagingConnectionNotifications consumerEventHandlers = null)
```

When I make the call I’m getting the error: 

> A suitable constructor for type 'Messaging.RabbitMq.RabbitMqConnection' could not be located. Ensure the type is concrete and services are registered for all parameters of a public constructor.

From doing some debugging, it seems that the call to CreateInstance is trying to find a constructor that has three parameters, while mine has five.  I’m expecting the first two parameters of my constructor to be resolved by a call to the DI Service Provider.

That doesn’t seem to be happening.  Is my understanding of how ActivatorUtilities.CreateInstance works incorrect? 

Would it be possible to get an example of how to use ActivatorUtilities.CreateInstance when some of the constructor parameters are expected to be resolved by DI?  Would it also be possible to get an example of how to use the CreateFactory method?

Thanks

-marc

**Comment by mhsimkin** [comment]
Guys, is anyone able to answer this query for me? Thanks.

**Comment by glennc** [comment]
@pakrym I would've thought this example works. Can you tell me what about DI makes it fail?

**Comment by pakrym** [comment]
Looks like it should work too, @mhsimkin can you provide a simplified runnable repro so I can look into it?

**Comment by mhsimkin** [comment]
@pakrym I can.  It will take me a day or two before I can get to it.

**Comment by pakrym** [comment]
Hi. We're closing this issue as no response or updates have been provided in a timely manner and we have been unable to reproduce it. If you have more details and are encountering this issue please add a new reply and re-open the issue.

**Comment by mhsimkin** [comment]
[DI-Issue.zip](https://github.com/aspnet/Extensions/files/3403493/DI-Issue.zip)
My apologies for taking so long to get back to this issue.  My priorities changed, which resulted in this issue/project being placed on the shelf.

I have attached a Zip file with a runnable solution that demonstrates the issue.

Inside the constructor of RabbitMqClientFactory I set up a local instance of the DI container and add some transient services.  When I call the GetMessagingConnection method of my factory class the error:

A suitable constructor for type 'DI_Issue.IRabbitMqConnection' could not be located. Ensure the type is concrete and services are registered for all parameters of a public constructor."

Is thrown.

The values of the parameters that need to be passed to the constructor of the RabbitMqConnection will not be known until runtime.

I’m trying to understand what I’m doing wrong.  As well as the best way to resolve a service and pass it additional parameters that are not known to the service resolver.

Thanks

-marc
